### PR TITLE
Thread safe AnalysisCallback 405

### PR DESCRIPTION
**Pull request by muxanick**
Small contributions from Dublin Scala Sprees:
I'm working on #405 
Unit test is complicated, we discussed that it would not be included to this PR

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?zinc=pull/409/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?zinc=pull/409/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?zinc=pull/409/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/179/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.x-pull/409/head-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/179/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@34945af6314dee3c9ea55a76c882007f44819509
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/409">pull/409/head       </A> | sbt/zinc@7be7b1dd301c7094f601506a03fbd686e34fb4b0
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@62004b28dfcdc33273ebb07a563e78178f6eb21d
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@04193716da1d7454d1d7c5935b396ebeefc8c2d4
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@cba6f83447b84cc1e967f25a930d4f411f6f0664
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=pull/409/head&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false">(restart)</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/180/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.x-pull/409/head-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/180/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@34945af6314dee3c9ea55a76c882007f44819509
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/409">pull/409/head       </A> | sbt/zinc@7be7b1dd301c7094f601506a03fbd686e34fb4b0
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@62004b28dfcdc33273ebb07a563e78178f6eb21d
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@04193716da1d7454d1d7c5935b396ebeefc8c2d4
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@cba6f83447b84cc1e967f25a930d4f411f6f0664
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=pull/409/head&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false">(restart)</A>

File: internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala [hunk 1]
```diff
@@ -141,40 +144,62 @@ private final class AnalysisCallback(
 
   import collection.mutable.{ HashMap, HashSet, ListBuffer, Map, Set }
```
Comment (dragos) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala:145]: We should probably remove unused imports from this list

File: internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala [hunk 2]
```diff
@@ -141,40 +144,62 @@ private final class AnalysisCallback(
...
-  private[this] val classPublicNameHashes = new HashMap[String, Array[NameHash]]
-  private[this] val objectPublicNameHashes = new HashMap[String, Array[NameHash]]
-  private[this] val usedNames = new HashMap[String, Set[UsedName]]
-  private[this] val unreporteds = new HashMap[File, ListBuffer[Problem]]
-  private[this] val reporteds = new HashMap[File, ListBuffer[Problem]]
-  private[this] val mainClasses = new HashMap[File, ListBuffer[String]]
-  private[this] val binaryDeps = new HashMap[File, Set[File]]
+  private[this] class SyncHashMap[K, V] extends mutable.Map[K, V] {
+
+    private val underline = new java.util.concurrent.ConcurrentHashMap[K, V]
```
Comment (dragos) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala:149]: typo: `underlying`

File: internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala [hunk 3]
```diff
@@ -141,40 +144,62 @@ private final class AnalysisCallback(
...
-  private[this] val binaryClassName = new HashMap[File, String]
+  private[this] val extSrcDeps = new SyncHashMap[String, Set[ExternalDependency]]
+  private[this] val binaryClassName = new SyncHashMap[File, String]
   // source files containing a macro def.
-  private[this] val macroClasses = Set[String]()
+  private[this] val macroClasses = new SyncHashMap[String, Boolean]()
 
-  private def add[A, B](map: Map[A, Set[B]], a: A, b: B): Unit = {
+  private def add[A, B](map: SyncHashMap[A, Set[B]], a: A, b: B): Unit = {
     map.getOrElseUpdate(a, new HashSet[B]) += b
```
Comment (dragos) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala:195]: The default `HashSet` should also be replaced by a thread-safe version.

**Comment by dragos** [comment]
I think it would be simpler to use the Scala `TrieMap` implementation and `ConcurrentHashMap.newKeySet` instead of wrapping the mutable `HashMap`.

**Comment by muxanick** [comment]
@dragos @dotta should I close this PR? seems Dotta's solution is better than mine

**Comment by dotta** [comment]
Hey @muxanick, I'm sorry I didn't share our implementation sooner: I really didn't know there was a ticket for having a thread-safe AnalysisCallback :-) About which PR to keep open, I'll defer to @jvican (in fact, I'm really ok if you guys decide to keep this PR open and cherrypick the relevant bits from my PR). Let me know if I can help with anything!

**Comment by jvican** [comment]
Hey @muxanick, thanks for this contribution! Yes, I think it'll be better to review/merge @dotta's PR, since it is more complete. Thank you for getting the time to make this though, I appreciate it :smile:.

**Pull request closed without merging**